src/ray/raylet/node_manager.cc

```diff
@@ -524,34 +524,23 @@ void NodeManager::ProcessNewClient(LocalClientConnection &client) {
   client.ProcessMessages();
 }
 
-void NodeManager::DispatchTasks(const std::list<Task> &ready_tasks) {
-  // Return if there are no tasks to schedule.
-  if (ready_tasks.empty()) {
-    return;
-  }
-
-  // Remember ids of the task we need to remove from ready queue.
-  std::unordered_set<TaskID> removed_task_ids = {};
-
-  for (const auto &task : ready_tasks) {
-    const auto &task_resources = task.GetTaskSpecification().GetRequiredResources();
-    if (!local_available_resources_.Contains(task_resources)) {
-      // Not enough local resources for this task right now, skip this task.
-      // TODO(rkn): We should always skip node managers that have 0 CPUs.
-      continue;
-    }
-    // We have enough resources for this task. Assign task.
-    if (AssignTask(task)) {
-      // We were successful in assigning this task on a local worker, so
-      // remember to remove it from ready queue. If for some reason the
-      // scheduling of this task fails later, we will add it back to the
-      // ready queue.
-      removed_task_ids.insert(task.GetTaskSpecification().TaskId());
+void NodeManager::DispatchTasks() {
+  std::unordered_set<TaskID> removed_task_ids;
+  for (auto &it : local_queues_.GetReadyQueue().GetTasksWithResources()) {
```

Comment:
While it's true that `unordered_map` doesn't provide ordering guarantees, I would've still preferred to randomize or round robin access to resource shape buckets. This combined with `std::deque` within each bucket might be good enough for approximate FIFO.

Reply:
This PR is not attempting to fix the starvation issue, that should be done in a different PR.

Reply:
Use `const auto &` here?